--- gdb-7.7.orig/sim/common/cgen-ops.h	2014-04-30 09:13:57.000000000 +0000
+++ gdb-7.7/sim/common/cgen-ops.h	2014-04-30 09:13:48.000000000 +0000
@@ -26,7 +26,7 @@
 
 #if defined (__GNUC__) && ! defined (SEMOPS_DEFINE_INLINE)
 #define SEMOPS_DEFINE_INLINE
-#define SEMOPS_INLINE extern inline
+#define SEMOPS_INLINE static inline
 #else
 #define SEMOPS_INLINE
 #endif
--- gdb-7.7.orig/sim/common/cgen-mem.h	2014-04-30 10:37:07.000000000 +0000
+++ gdb-7.7/sim/common/cgen-mem.h	2014-04-30 10:37:36.000000000 +0000
@@ -23,7 +23,7 @@
 #ifdef MEMOPS_DEFINE_INLINE
 #define MEMOPS_INLINE
 #else
-#define MEMOPS_INLINE extern inline
+#define MEMOPS_INLINE static inline
 #endif
 
 /* Integer memory read support.
