diff -ruw gdb-7.11.orig/sim/erc32/sis.h gdb-7.11/sim/erc32/sis.h
--- gdb-7.11.orig/sim/erc32/sis.h	2016-03-15 10:24:35.934052479 +1100
+++ gdb-7.11/sim/erc32/sis.h	2016-03-16 13:52:51.988554970 +1100
@@ -20,7 +20,7 @@
 #include <sim-config.h>
 #include <stdint.h>
 
-#if HOST_BYTE_ORDER == BIG_ENDIAN
+#ifdef WORDS_BIGENDIAN
 #define HOST_BIG_ENDIAN
 #define EBT 0
 #else
Only in gdb-7.11/sim/erc32: sis.h~
