Index: gcc/config/microblaze/rtems.h
===================================================================
--- gcc/config/microblaze/rtems.h       (revision 196028)
+++ gcc/config/microblaze/rtems.h       (working copy)
@@ -23,3 +23,15 @@
   builtin_define( "__rtems__" );               \
   builtin_assert( "system=rtems" );            \
 } while (0)
+
+/* Use the default */
+#undef LINK_GCC_C_SEQUENCE_SPEC
+
+/* Extra switches sometimes passed to the linker.  */
+/* -xl-mode-xmdstub translated to -Zxl-mode-xmdstub -- deprecated. */
+/* RTEMS: Remove use of xilinx.ld but keep other parts for compatibility */
+#undef LINK_SPEC
+#define LINK_SPEC "%{shared:-shared} -N -relax \
+  %{Zxl-mode-xmdstub:-defsym _TEXT_START_ADDR=0x800} \
+  %{mxl-mode-xmdstub:-defsym _TEXT_START_ADDR=0x800} \
+  %{mxl-gp-opt:%{G*}} %{!mxl-gp-opt: -G 0}"
