diff --git a/libgcc/config.host b/libgcc/config.host
index 60340a6..2921ca8 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -714,7 +714,7 @@ mcore-*-elf)
 microblaze*-linux*)
 	tmake_file="${tmake_file} microblaze/t-microblaze t-fdpbit t-slibgcc-libgcc"
 	;;
-microblaze*-*-elf)
+microblaze*-*-elf | microblaze*-*-rtems*)
 	tmake_file="${tmake_file} microblaze/t-microblaze t-fdpbit"
 	extra_parts="$extra_parts crtbeginS.o crtendS.o crtbeginT.o crti.o crtn.o"
 	;;
